app/src/main/java/org/mozilla/fenix/home/recentvisits/RecentVisitsFeature.kt

```diff
@@ -0,0 +1,131 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.home.recentvisits
+
+import androidx.annotation.VisibleForTesting
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.async
+import kotlinx.coroutines.launch
+import mozilla.components.browser.storage.sync.PlacesHistoryStorage
+import mozilla.components.concept.storage.HistoryHighlight
+import mozilla.components.concept.storage.HistoryHighlightWeights
+import mozilla.components.concept.storage.HistoryMetadata
+import mozilla.components.concept.storage.HistoryMetadataStorage
+import mozilla.components.support.base.feature.LifecycleAwareFeature
+import org.mozilla.fenix.home.HomeFragment
+import org.mozilla.fenix.home.HomeFragmentAction
+import org.mozilla.fenix.home.HomeFragmentStore
+import org.mozilla.fenix.home.recentvisits.RecentlyVisitedItem.RecentHistoryGroup
+import org.mozilla.fenix.home.recentvisits.RecentlyVisitedItem.RecentHistoryHighlight
+import kotlin.math.max
+
+@VisibleForTesting internal const val MAX_RESULTS_TOTAL = 9
+@VisibleForTesting internal const val MAX_RESULTS_OF_EACH_TYPE = 6
+@VisibleForTesting internal const val MIN_VIEW_TIME_OF_HIGHLIGHT = 10.0
+@VisibleForTesting internal const val MIN_FREQUENCY_OF_HIGHLIGHT = 4.0
+
+/**
+ * View-bound feature that retrieves a list of [HistoryHighlight]s and [HistoryMetadata] items
+ * which will be mapped to [RecentlyVisitedItem]s and then dispatched to [HomeFragmentStore]
+ * to be displayed on the homescreen.
+ *
+ * @param homeStore The [HomeFragmentStore] that holds the state of the [HomeFragment].
+ * @param historyMetadataStorage The storage that manages [HistoryMetadata].
+ * @param historyHighlightsStorage The storage that manages [PlacesHistoryStorage].
+ * @param scope The [CoroutineScope] used for IO operations related to querying history
+ * and then for dispatching updates.
+ * @param ioDispatcher The [CoroutineDispatcher] for performing read/write operations.
+ */
+class RecentVisitsFeature(
+    private val homeStore: HomeFragmentStore,
+    private val historyMetadataStorage: HistoryMetadataStorage,
+    private val historyHighlightsStorage: Lazy<PlacesHistoryStorage>,
+    private val scope: CoroutineScope,
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO,
+) : LifecycleAwareFeature {
+
+    private var job: Job? = null
+
+    override fun start() {
+        job = scope.launch(ioDispatcher) {
```

Comment:
Probably prefer to have @csadilek review this portion of the PR

Reply:
Sure! Thank you for the quick review!